### ISSUE TITLE
OneDrive authentication for syncing, opening a file fails in Firefox

### ISSUE THREAD

**Issue by thenaterhood** [keeweb/keeweb#1713]
**Describe the bug**
This is a new issue, as I've been successfully syncing with OneDrive for quite a while. I don't know if this is an issue with changes in Firefox. I don't even know if this is within the power of Keeweb to fix.

Keeweb in Firefox is failing to authenticate with OneDrive because of a blocked CORS request. From the Firefox console:

**To Reproduce**
1. Try to load a file from OneDrive in Firefox 85.0.2
2. The Microsoft oauth popup should appear. Grant permission if it asks.
3. The process will fail

This happens with a previously configured and authenticated setup (token expired and it wants to re-auth maybe?) as well as a brand new setup (tried on beta.keeweb which I've never used before)

**Expected behavior**
OneDrive can successfully authenticate and sync.

**Screenshots**

**Environment**
KeeWeb v1.16.7 (42b5b00, 2020-12-31)
Environment: web
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:85.0) Gecko/20100101 Firefox/85.

**Kdbx File**
N/A

**Logs**

```
2021-02-13T00:22:48.893Z [sync:thenaterhood] Sync started onedrive /drive/root:/thenaterhood.kdbx 
Object {  }
app.keeweb.info:1:387074
2021-02-13T00:22:48.908Z [sync:thenaterhood] Stat file app.keeweb.info:1:387074
2021-02-13T00:22:49.082Z [storage-onedrive] OAuth: popup opened app.keeweb.info:1:386953
2021-02-13T00:22:50.097Z [storage-onedrive] OAuth code received app.keeweb.info:1:386953
2021-02-13T00:22:50.098Z [storage-onedrive] HTTP request POST https://login.microsoftonline.com/common/oauth2/v2.0/token app.keeweb.info:1:387074
2021-02-13T00:22:50.538Z [storage-onedrive] Error exchanging OAuth code network error app.keeweb.info:1:387321
2021-02-13T00:22:50.538Z [sync:thenaterhood] Stat error, not dirty OAuth code exchange error: network error app.keeweb.info:1:387074
2021-02-13T00:22:50.538Z [sync:thenaterhood] Sync finished OAuth code exchange error: network error app.keeweb.info:1:387074
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://login.microsoftonline.com/common/oauth2/v2.0/token. (Reason: CORS request did not succeed).

Failed to load ‘https://login.microsoftonline.com/common/oauth2/v2.0/token’. A ServiceWorker passed a promise to FetchEvent.respondWith() that rejected with ‘TypeError: NetworkError when attempting to fetch resource.’.
```

**Comment by antelle** [comment]
Hi! Are you using a personal or business account?

**Comment by thenaterhood** [comment]
I'm using a person account.

**Comment by antelle** [comment]
Just checked, it doesn't work in all browsers. I assume, Microsoft changed some settings, will look into it.

**Comment by antelle** [comment]
I'm afraid, we may have to switch to SPA application type in OneDrive, which means it will result in an authentication pop-up being displayed every time 😢

**Comment by Muscipular** [comment]
chrome is same,  console reported:
```
Access to fetch at 'https://login.microsoftonline.com/common/oauth2/v2.0/token' from origin 'https://app.keeweb.info' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Comment by jortfal** [comment]
Same error in Microsoft  #Edge:

_Access to fetch at 'https://login.microsoftonline.com/common/oauth2/v2.0/token' from origin 'https://app.keeweb.info' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled._

Endpoint https://login.microsoftonline.com/common/oauth2/authorize does not support CORS

@antelle as you indicate the solution switch to SPA:

https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2254#issuecomment-689884380

**Comment by antelle** [comment]
@jortfal yeah I'll test something else and if nothing helps, we'll switch to SPA (which has some drawbacks).

**Comment by obnoxiousnerd** [comment]
Can we work around this by using a CORS proxy?

**Comment by antelle** [comment]
No, you don't want this proxy to have access to all your data in OneDrive

**Comment by obnoxiousnerd** [comment]
@antelle Yeah, that perfectly makes sense, considering I didn't explained further. Anyway I tried with a proxy but no luck. I guess then shifting to MSAL 2.x will work.

> I'm afraid, we may have to switch to SPA application type in OneDrive, which means it will result in an authentication pop-up being displayed every time 😢

IMO, that's pretty strange. I've never seen that kind of behavior. Can you try switching to SPA type, test and tell us?

BTW, you are doing a pretty good job by making this web app. I support you and maybe I'll also try my hands on the code :)

**Comment by antelle** [comment]
I tried on an example, and it looks like there's no way to make tokens last long. It's also said in the [Microsoft's doc](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#refresh-the-access-token):

> Single page apps, however, get a token with a 24 hour lifetime, requiring a new authentication every day.

> I've never seen that kind of behavior.

Probably because most of apps are not SPA.

**Comment by antelle** [comment]
I can't find any other way of authentication, so I'll make a release with this next week. Desktop apps should continue working as is, while webapp will prompt for authentication from time to time.

**Comment by antelle** [comment]
I made the change, however they also mention invisible iframes in the docs, that they can be used to obtain a new token silently, this might worth trying.

**Comment by antelle** [comment]
Now OneDrive is working, however it will ask to re-authenticate once a day. I'll look into this iframe thing in the next release.
